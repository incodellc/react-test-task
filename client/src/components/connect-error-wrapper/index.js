import ConnectErrorWrapper from './ConnectErrorWrapper';
export default ConnectErrorWrapper;