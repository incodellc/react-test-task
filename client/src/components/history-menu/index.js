import HistoryMenu from './HistoryMenu';
export default HistoryMenu;