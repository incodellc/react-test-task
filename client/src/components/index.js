import InfoBar from './info-bar';
import Layout from './layout';
import ConnectErrorWrapper from './connect-error-wrapper';
import TickerList from './ticker-list';
import TickerListItem from './ticker-list-item';

export { InfoBar, Layout, ConnectErrorWrapper, TickerList, TickerListItem };