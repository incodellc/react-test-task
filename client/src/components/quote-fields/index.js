import QuoteFields from './QuoteFields';
export default QuoteFields;