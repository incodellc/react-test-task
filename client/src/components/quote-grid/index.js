import QuotesGrid from './QuotesGrid';
export default QuotesGrid;