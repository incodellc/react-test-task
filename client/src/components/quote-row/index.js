import QuoteRow from './QuoteRow';
export default QuoteRow;