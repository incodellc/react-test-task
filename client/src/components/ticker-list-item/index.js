import TickerListItem from './TickerListItem';
export default TickerListItem;