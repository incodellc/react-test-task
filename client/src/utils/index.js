import moment from 'moment';
export { quotesExample, tickers, fields, historyLengthOptions } from './presets';

export const rewriteObjectProps = (prevObject, newPropsValues) => {
    return {
        ...prevObject,
        ...newPropsValues
    };
};

export const sortByKey = (arrayOfObjects, sortParams) => {
    let { target, isDate, reverse } = sortParams;
    let data = arrayOfObjects.slice();

    if (target === null) {
        return data;
    }

    const compareByKey = (key, isDate = false, isDesc = false) => {
        let orderFlag = isDesc ? -1 : 1;
        return (a, b) => {
            let valA = typeof a[key] === 'string' && isNaN(a[key]) ? a[key].toUpperCase() : typeof +a[key] ===  'number' ? +a[key] : '';
            let valB = typeof b[key] === 'string' && isNaN(a[key]) ? b[key].toUpperCase() : typeof +b[key] ===  'number' ? +b[key] : '';
            if (isDate) {
                valA = valA === '' ? moment(+valA) : moment(valA, 'DD-MM-YYYY HH:mm');
                valB = valB === '' ? moment(+valB) : moment(valB, 'DD-MM-YYYY HH:mm');
            }
            let result = (valA < valB) ? -1 : (valA > valB) ? 1 : 0;
            return result * orderFlag;
        };
    };

    return data.sort(compareByKey(target, isDate, reverse));
};
