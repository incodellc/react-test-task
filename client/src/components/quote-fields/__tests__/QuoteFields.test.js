import React from 'react';
import { fields } from '../../../utils';
import { renderWithRedux } from '../../../utils/test-utils';
import { QuoteFields } from '../../../components';
import { cleanup, fireEvent } from '@testing-library/react';

describe('Check QuoteFields component', () => {
    afterEach(cleanup);

    it('component renders correctly', () => {
        const { asFragment } = renderWithRedux(<QuoteFields />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('component should render row of cells witch must match the order, aliases and number of fields', () => {
        const { getByTestId, getAllByTestId } = renderWithRedux(<QuoteFields />);
        const cells = getAllByTestId('quote-fields-cell');
        expect(getByTestId('quote-fields')).toContainElement(cells[0]);
        expect(cells.length).toEqual(fields.length);
        fields.forEach((field, index) => {
            expect(cells[index]).toHaveTextContent(field.alias);
        });
    });
    
    it('component should receive updated state and react correctly after click event fired', () => {
        const { getByText } = renderWithRedux(<QuoteFields />);
        expect(getByText('Ticker')).not.toHaveClass('sorted');
        fireEvent.click(getByText('Ticker'));
        expect(getByText('Ticker')).toHaveClass('sorted ascending');
        fireEvent.click(getByText('Ticker'));
        expect(getByText('Ticker')).toHaveClass('sorted descending');
        fireEvent.click(getByText('Ticker'));
        expect(getByText('Ticker')).toHaveClass('sorted ascending');
        fireEvent.click(getByText('Price'));
        expect(getByText('Ticker')).not.toHaveClass('sorted');
        expect(getByText('Price')).toHaveClass('sorted ascending');
    });
});