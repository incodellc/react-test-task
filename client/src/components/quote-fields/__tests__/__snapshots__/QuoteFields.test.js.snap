// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check QuoteFields component component renders correctly 1`] = `
<DocumentFragment>
  <table
    class="ui celled fixed sortable table quoteFields"
    data-testid="quote-fields"
  >
    <thead
      class=""
    >
      <tr
        class=""
      >
        <th
          class="disabled"
          data-testid="quote-fields-cell"
        >
          Exchange
        </th>
        <th
          class=""
          data-testid="quote-fields-cell"
        >
          Ticker
        </th>
        <th
          class=""
          data-testid="quote-fields-cell"
        >
          Price
        </th>
        <th
          class=""
          data-testid="quote-fields-cell"
        >
          Change
        </th>
        <th
          class=""
          data-testid="quote-fields-cell"
        >
          Change %
        </th>
        <th
          class=""
          data-testid="quote-fields-cell"
        >
          Dividend
        </th>
        <th
          class=""
          data-testid="quote-fields-cell"
        >
          Yield
        </th>
        <th
          class=""
          data-testid="quote-fields-cell"
        >
          Last trade time
        </th>
      </tr>
    </thead>
  </table>
</DocumentFragment>
`;
